rotkehlchen/chain/aggregator.py

```diff
@@ -1339,19 +1359,57 @@ def add_accounts_to_all_evm(
         """
-        filtered_accounts = self.filter_active_evm_addresses(accounts)
-        added_accounts = []
-        for chain, account in filtered_accounts:
-            try:
-                self.modify_blockchain_accounts(
-                    blockchain=chain,
-                    accounts=[account],
-                    append_or_remove='append',
+        added_accounts: list[tuple[SUPPORTED_EVM_CHAINS, ChecksumEvmAddress]] = []
+        # Distinguish between contracts and EOAs
+        for account in accounts:
+            if self.is_contract(account, SupportedBlockchain.ETHEREUM):
+                added_chains = self.track_evm_address(account, [SupportedBlockchain.ETHEREUM])
+                if len(added_chains) == 1:  # Is always either 1 or 0 since is only for ethereum
+                    added_accounts.append((SupportedBlockchain.ETHEREUM, account))
+            else:
+                added_accounts += self.check_chains_and_add_accounts(
+                    account=account,
+                    chains=list(typing.get_args(SUPPORTED_EVM_CHAINS)),
                 )
-            except InputError:
-                log.debug(f'Not adding {account} to {chain} since it already exists')
+
+        return added_accounts
+
+    def detect_evm_accounts(
+            self,
+            progress_handler: Optional['ProgressUpdater'] = None,
+    ) -> list[tuple[SUPPORTED_EVM_CHAINS, ChecksumEvmAddress]]:
+        current_accounts: dict[ChecksumEvmAddress, list[SUPPORTED_EVM_CHAINS]] = defaultdict(list)
+        chain: SUPPORTED_EVM_CHAINS
+        for chain in typing.get_args(SUPPORTED_EVM_CHAINS):
+            chain_accounts = self.accounts.get(chain)
+            for account in chain_accounts:
+                current_accounts[account].append(chain)
+
+        added_accounts: list[tuple[SUPPORTED_EVM_CHAINS, ChecksumEvmAddress]] = []
+        for account, chains in current_accounts.items():
+            if progress_handler is not None:
+                progress_handler.new_step(f'Checking {account} EVM chain activity')
+
+            if self.is_contract(account, chains[0]):
+                continue  # do not check contracts
+
+            chains_to_check = list(set(typing.get_args(SUPPORTED_EVM_CHAINS)) - set(chains))
+            if len(chains_to_check) == 0:
                 continue
-            added_accounts.append((chain, account))
+
+            added_accounts += self.check_chains_and_add_accounts(account, chains_to_check)
+
+        if progress_handler is not None:
+            progress_handler.new_step('Potentially write migrated addresses to the DB')
+
+        with self.database.user_write() as write_cursor:
+            self.database.add_blockchain_accounts(
+                write_cursor=write_cursor,
+                account_data=[
+                    BlockchainAccountData(chain, account)
+                    for chain, account in added_accounts
+                ],
+            )
```

Comment:
keep the comment since it helps later to remember why things were done

```suggestion
        with self.database.user_write() as write_cursor:
            self.database.add_blockchain_accounts(
                write_cursor=write_cursor,
                account_data=[
                    BlockchainAccountData(chain, account)
                    for chain, account in added_accounts
                ],    # not duplicating label and tags as it's chain specific
            )
```